File: code/day5.jl
# Advent of Code Day 5 Julia Solution

function findSeats(file)
    codes = readlines(file)
    [toDecimal(code) for code in codes]
end

function toDecimal(x::AbstractString)
    binary = replace(replace(x, r"F|L" => "0"), r"B|R" => "1")
    count, pow = 0, length(binary)
    for i in 1:(pow)
        count += parse(Int, binary[i]) * 2^(pow - i)
    end
    count
end

seats = findSeats("data/day5")

# puzzle 1
maximum(seats)

# puzzle 2
sort!(seats)
for seat in seats
    if seats[seat + 1] - seats[seat] == 2
        println(seats[seat] + 1)
        break
    end
end

File: code/day5.R
# Day 5 Advent of Code R Solution
# This is the binary solution, but this is a beat of cheat because I first solved without binary.
find_seat <- function(file) {
  codes <- scan(file, what = character())
  s <- gsub("B|R", "1", gsub("F|L", "0", codes))
  row <- strtoi(substr(s, 1, 7), 2)
  col <- strtoi(substr(s, 8, 10), 2)
  row * 8 + col
}

ids <- sort(find_seat("data/day5"))

# puzzle 1 
max(ids)

# puzzle 2
for (i in 1:(length(ids) - 1)) {
  if (ids[i + 1] - ids[i] == 2) {
    print(ids[i] + 1)
  }
}
